state-synchronizer/src/coordinator.rs

```diff
@@ -1043,15 +1051,24 @@ impl<T: ExecutorProxyTrait> SyncCoordinator<T> {
         // Optimistically fetch the next chunk.
         let new_version =
             self.local_state.highest_version_in_local_storage() + txn_list_with_proof.len() as u64;
-        // The epoch in the optimistic request should be the next epoch if the current chunk
+        // The epoch in the optimistic request (= new_epoch) should be the next epoch if the current chunk
         // is the last one in its epoch.
-        let new_epoch = end_of_epoch_li
-            .as_ref()
-            .map_or(self.local_state.epoch(), |li| {
-                if li.ledger_info().version() == new_version {
-                    self.local_state.epoch() + 1
+        // Also perform some verifications on the end-of-epoch-li passed to execution:
+        // * Verify end-of-epoch-li's version corresponds to end-of-chunk version before
+        // passing it to execution
+        // * Executor expects that when it is passed an end-of-epoch LI, it is going to execute/commit
+        // transactions leading up to that end-of-epoch LI
+        let (new_epoch, end_of_epoch_li) = end_of_epoch_li
+            .map(|li| {
+                // verify end-of-epoch-li against local state
+                self.local_state.trusted_epoch.verify(&li).map(|_| li)
+            })
+            .transpose()?
```

Comment:
I thought the plan was to take the li and do the optimistic follow up here

down below where we have

`self.waypoint.verify(`

we could do

end_of_epoch_li.map(|li| self...).transpose()?;